### PR TITLE
refactor(bitbucket-server): update runtime pr list

### PR DESCRIPTION
**Pull request by RahulGautamSingh**
## Changes
Update the pr list stored at runtime with new/updated info when a pr is updated.

## Context
Ref: https://github.com/renovatebot/renovate/issues/32040

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [x] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/platform/bitbucket-server/index.spec.ts [hunk 1]
```diff
@@ -1694,16 +1694,22 @@ describe('modules/platform/bitbucket-server/index', () => {
...
-              targetBranch: 'new_base',
-            }),
-          ).toResolve();
+          await bitbucket.updatePr({
+            number: 5,
+            prTitle: 'title',
+            prBody: 'body',
+            targetBranch: 'new_base',
+          });
+          expect(1).toBe(1);
```
Comment (viceice) [inline lib/modules/platform/bitbucket-server/index.spec.ts:1712]: why this change? I would prefer the previous one

File: lib/modules/platform/bitbucket-server/index.ts [hunk 2]
```diff
@@ -1011,22 +1011,25 @@ export async function updatePr({
       };
     }
 
-    const { body: updatedPr } = await bitbucketServerHttp.putJson<{
-      version: number;
-      state: string;
-    }>(
+    const { body: updatedPrRes } = await bitbucketServerHttp.putJson<
+      BbsRestPr & {
```
Comment (viceice) [inline lib/modules/platform/bitbucket-server/index.ts:1015]: doesn't this interface already have version and state? 🤔

File: lib/modules/platform/bitbucket-server/index.ts [hunk 3]
```diff
@@ -1011,22 +1011,25 @@ export async function updatePr({
       };
     }
 
-    const { body: updatedPr } = await bitbucketServerHttp.putJson<{
-      version: number;
-      state: string;
-    }>(
+    const { body: updatedPrRes } = await bitbucketServerHttp.putJson<
```
Comment (viceice) [inline lib/modules/platform/bitbucket-server/index.ts:1014]: can you avoid the remaining in this pr?

File: lib/modules/platform/bitbucket-server/index.spec.ts (hunk 1 above)
Comment (RahulGautamSingh) [reply lib/modules/platform/bitbucket-server/index.spec.ts:1712]: Sorry, got left :)

File: lib/modules/platform/bitbucket-server/index.ts (hunk 3 above)
Comment (RahulGautamSingh) [reply lib/modules/platform/bitbucket-server/index.ts:1014]: Didn't get it

Comment (viceice) [reply lib/modules/platform/bitbucket-server/index.ts]:
```suggestion
    const { body: updatedPr } = await bitbucketServerHttp.putJson<BbsRestPr>(
```

**Review by viceice — COMMENTED** [review]
see comment

File: lib/modules/platform/bitbucket-server/index.ts (hunk 3 above)
Comment (RahulGautamSingh) [reply lib/modules/platform/bitbucket-server/index.ts]: Why though?

File: lib/modules/platform/bitbucket-server/index.ts [hunk 4]
```diff
@@ -1011,22 +1011,20 @@ export async function updatePr({
       };
     }
 
-    const { body: updatedPr } = await bitbucketServerHttp.putJson<{
-      version: number;
-      state: string;
-    }>(
+    const { body: updatedPr } = await bitbucketServerHttp.putJson<BbsRestPr>(
```
Comment (viceice) [inline lib/modules/platform/bitbucket-server/index.ts:1014]:
```suggestion
    const { body: updatedPr } = await bitbucketServerHttp.putJson<BbsRestPr & {
        version: number;
      }>(
```

wonder why `version` is optional. 🤔  

we should switch to zod schema as soon as possible

File: lib/modules/platform/bitbucket-server/index.ts [hunk 5]
```diff
@@ -1011,22 +1011,20 @@ export async function updatePr({
...
       `./rest/api/1.0/projects/${config.projectKey}/repos/${config.repositorySlug}/pull-requests/${prNo}`,
       { body },
     );
 
-    updatePrVersion(prNo, updatedPr.version);
+    updatePrVersion(prNo, updatedPr.version!);
```
Comment (viceice) [inline lib/modules/platform/bitbucket-server/index.ts:1019]:
```suggestion
    updatePrVersion(prNo, updatedPr.version);
```

File: lib/modules/platform/bitbucket-server/index.ts [hunk 6]
```diff
@@ -1011,22 +1011,20 @@ export async function updatePr({
...
 
     const currentState = updatedPr.state;
     // TODO #22198
     const newState = {
       ['open']: 'OPEN',
       ['closed']: 'DECLINED',
     }[state!];
+    let finalState: string = currentState;
```
Comment (viceice) [inline lib/modules/platform/bitbucket-server/index.ts:1027]: why do we need this `string` cast?

File: lib/modules/platform/bitbucket-server/index.ts [hunk 7]
```diff
@@ -1040,8 +1038,23 @@ export async function updatePr({
...
     }
+
+    if (config.prList) {
+      const bbsPr = utils.prInfo(updatedPr);
+      const existingIndex = config.prList.findIndex(
+        (item) => item.number === prNo,
+      );
+      // istanbul ignore if: should never happen
+      if (existingIndex === -1) {
+        config.prList.push({ ...bbsPr, state: finalState });
```
Comment (viceice) [inline lib/modules/platform/bitbucket-server/index.ts:1053]: log an warn / error so someone sees if it happens on runtime

File: lib/modules/platform/bitbucket-server/index.ts (hunk 4 above)
Comment (RahulGautamSingh) [reply lib/modules/platform/bitbucket-server/index.ts:1014]: its optional in the `BbsRestPr` interface..hence :) Should be changed, is it okay to do in this pr?

File: lib/modules/platform/bitbucket-server/index.ts (hunk 6 above)
Comment (RahulGautamSingh) [reply lib/modules/platform/bitbucket-server/index.ts]: Nice catch! Thanks... It was a wrong cast and wrong value too.

**Review by viceice — APPROVED** [review]
